File: src/main/java/com/ocunha/api/RestaurantsApiImpl.java
package com.ocunha.api;

import com.ocunha.domain.restaurant.RestaurantService;
import com.ocunha.domain.restaurant.model.RestaurantSearchParams;
import com.ocunha.domain.restaurant.model.model.Restaurant;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequiredArgsConstructor
public class RestaurantsApiImpl implements RestaurantsApi {

    private final RestaurantService restaurantService;

    @Override
    public ResponseEntity<List<RestaurantResource>> getRestaurants(String name, Integer rating, Integer distance, Integer price, String cuisine) {
        RestaurantSearchParams searchParams = RestaurantSearchParams.builder()
                .name(name)
                .rating(rating)
                .distance(distance)
                .price(price)
                .cuisine(cuisine)
                .build();

        return ResponseEntity.ok(restaurantService.findRestaurants(searchParams)
                .stream().map(this::getRestaurantResource).toList());
    }

    private RestaurantResource getRestaurantResource(Restaurant restaurant) {
        return RestaurantResource.builder()
                .name(restaurant.name())
                .cuisine(restaurant.cuisine())
                .price(String.valueOf(restaurant.price()))
                .distance(String.valueOf(restaurant.distance()))
                .rating(String.valueOf(restaurant.rating()))
                .build();
    }


}

File: src/main/kotlin/com/ocunha/api/RestaurantsApiImpl.kt
package com.ocunha.api

import com.ocunha.domain.restaurant.RestaurantService
import com.ocunha.domain.restaurant.model.Restaurant
import com.ocunha.domain.restaurant.model.RestaurantSearchParams
import com.ocunha.spec.api.RestaurantResource
import com.ocunha.spec.api.RestaurantsApi
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.RestController


@RestController
class RestaurantsApiImpl(private val restaurantService: RestaurantService) : RestaurantsApi {

    override fun getRestaurants(
        name: String?,
        rating: Int?,
        distance: Int?,
        price: Int?,
        cuisine: String?
    ): ResponseEntity<List<RestaurantResource>> {
        val restaurants = restaurantService.findRestaurants(
            RestaurantSearchParams.Builder()
                .name(name)
                .rating(rating)
                .distance(distance)
                .price(price)
                .cuisine(cuisine)
                .build()
        )
        return ResponseEntity.ok(restaurants?.map { restaurant -> getRestaurantResource(restaurant) });
    }

    private fun getRestaurantResource(restaurant: Restaurant?): RestaurantResource {
        val restaurantResource = RestaurantResource()
        restaurantResource.name = restaurant?.name
        restaurantResource.cuisine = restaurant?.cuisine
        restaurantResource.price = restaurant?.price
        restaurantResource.distance = restaurant?.distance
        restaurantResource.rating = restaurant?.rating
        return restaurantResource
    }
}